04-onion-routing.md

```diff
@@ -263,39 +264,108 @@ It is formatted according to the Type-Length-Value format defined in [BOLT #1](0
-    - MUST NOT include `short_channel_id`
-    - if the recipient provided `payment_secret`:
-      - MUST include `payment_data`
-      - MUST set `payment_secret` to the one provided
-      - MUST set `total_msat` to the total amount it will send
-    - if the recipient provided `payment_metadata`:
-      - MUST include `payment_metadata` with every HTLC
-      - MUST not apply any limits to the size of payment_metadata except the limits implied by the fixed onion size
+    - For every non-final node:
+      - MUST include `short_channel_id`
+      - MUST NOT include `payment_data`
+    - For the final node:
+      - MUST NOT include `short_channel_id`
+      - if the recipient provided `payment_secret`:
+        - MUST include `payment_data`
+        - MUST set `payment_secret` to the one provided
+        - MUST set `total_msat` to the total amount it will send
+      - if the recipient provided `payment_metadata`:
+        - MUST include `payment_metadata` with every HTLC
+        - MUST not apply any limits to the size of `payment_metadata` except the limits implied by the fixed onion size
 
 The reader:
-  - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
-  - if it is the final node:
-    - MUST treat `total_msat` as if it were equal to `amt_to_forward` if it
-      is not present.
+
+  - If `encrypted_recipient_data` is present:
+    - If `blinding_point` is set in the incoming `update_add_htlc`:
+      - MUST return an error if `current_blinding_point` is present.
+      - MUST use that `blinding_point` as the blinding point for decryption.
+    - Otherwise:
+      - MUST return an error if `current_blinding_point` is not present.
+      - MUST use that `current_blinding_point` as the blinding point for decryption.
+      - SHOULD add a random delay before returning errors.
+    - MUST return an error if `encrypted_recipient_data` does not decrypt using the
+      blinding point as described in [Route Blinding](#route-blinding).
+    - If `payment_constraints` is present:
+      - MUST return an error if:
+        - the expiry is greater than `encrypted_recipient_data.payment_constraints.max_cltv_expiry`.
+        - the amount is below `encrypted_recipient_data.payment_constraints.htlc_minimum_msat`.
+    - If `allowed_features` is missing:
+      - MUST process the message as if it were present and contained an empty array.
+    - If `allowed_features` is present:
+      - MUST return an error if:
+        - `encrypted_recipient_data.allowed_features.features` contains an unknown feature bit (even if it is odd).
+        - the payment uses a feature not included in `encrypted_recipient_data.allowed_features.features`.
+    - If it not the final node:
+      - MUST return an error if the payload contains other tlv fields than `encrypted_recipient_data` and `current_blinding_point`.
+      - MUST return an error if `encrypted_recipient_data` does not contain either `short_channel_id` or `next_node_id`.
```

Comment:
Answered in Spec PR. Interesting, I think LND may have strayed from making that decision based off its sphinx implementation encountering an all-zero onion HMAC, to instead checking for the presence of `short_channel_id` (perhaps these were functionally equivalent in practice). I expect this will need updating now that scid need not be set for blinded hops.